### ISSUE TITLE
Could we support ChannelShuffle? 

### ISSUE THREAD

**Issue by yinghai** [intel/mkl-dnn#241]
Hi folks, 
Shufflenet is a pretty popular model (https://arxiv.org/abs/1707.01083). And we tried to run in on MKL-DNN. The only thing that MKLDNN doesn't support is the ChannelShuffle primitive, so we ended up doing fallback to our our implementation and do the context switch. We noticed that MKLDNN's depthwise CNN really helps in this model but the performance is hampered by the context switch, which manifests as reordering of inputs at the conv op. Since channel shuffle is conceptually similar to reorder, is there any possibility that we can add support to that? It's will really benefit Shufflenet a lot. Here is a reference implementation of ChannelShuffle just to get the idea of what it is: https://github.com/pytorch/pytorch/blob/master/caffe2/operators/channel_shuffle_op.h#L14-L64

Thanks a lot.

**Comment by emfomenk** [comment]
Hi @yinghai,

Alas we don't have any plans to support shuffle operation at the moment.

If we are going to support that we need a separate primitive. If the performance problem mostly comes from the reorders from blocked layout to the native one and reverse reorder then having a simple implementation of shuffle in Intel MKL-DNN that would use our layout directly should help. We would be glad to get a PR on that feature. I also recall that IntelCaffe was interested in that feature...

---

Even though it is almost a reorder, I don't think we should use it cause of two reasons:
- reorder doesn't change the logical data, it just changes the physical layout. In other words the semantics of the data should stay the same. Shuffling on the other hand changes the semantics
- shuffling should not change the layout it self, so from reorder perspective there is nothing to do
- since we typically block over channels (`nChw16c`) that implementation would be much more complicated than just a reorder... (but that is an implementation detail, that is why i don't count that as a reason)
But that's all obvious probably.

**Comment by yinghai** [comment]
> If the performance problem mostly comes from the reorders from blocked layout to the native one and reverse reorder then having a simple implementation of shuffle in Intel MKL-DNN that would use our layout directly should help.

Channel shuffling is kinda like reordering if you think it as an primitive to reorder `ngkhw` to `nkghw` where `c=k*g`. But yes, the performance bottleneck is reorder from `nchw` to MKLDNN's blocked layout, at the input of the MKLDNN conv primitive. Is this something that is relatively easier than implementing ChannelShuffle? If I would like to help, any entry point that I can start looking?

**Comment by emfomenk** [comment]
Hi @yinghai,

Sorry for delay with the response.
I tried to put some quick intro to Intel MKL-DNN development but failed :) Will try to do that a little bit later.

Anyways, the simplest way to get started is by looking at some particular simple operation Intel MKL-DNN supports. Few examples might be: softmax, pooling, eltwise (ReLU, ...). Softmax is the simplest one cause we only have one C-based implementation (no jit or other fancy stuff). In order to get familiar with Intel MKL-DNN structure you should go over the following files:
```
$ grep softmax -Rl include src tests
include/mkldnn.h
...
```

Here is the full output of grep and short descriptions of the files:

| File | Description | Note |
| -- | -- | -- |
| include/mkldnn.h | C entry points | defines `mkldnn_softmax_forward_desc_init` and `mkldnn_softmax_backward_desc_init` |
| include/mkldnn.hpp | C++ entry points | defines `softmax_forward` and `softmax_backward` classes |
| include/mkldnn_types.h | C types | defines `softmax_desc_t` (in our term this is op(eration) descriptor -- the transparent part of future op *primitive* descriptor which is opaque |
| src/common/c_types_map.hpp | C2C++ types | mostly for convenience | 
| src/common/mkldnn_traits.hpp | aux traits | required for internal machinery |
| src/common/softmax.cpp | C entry points | the implementation for functions from `include/mkldnn.h` |
| src/common/softmax_pd.hpp | internal softmax primitive descriptor definition | _pd is a shortcut for primitive descriptor |
| src/common/mkldnn_debug.cpp | aux stuff for debugging | |
| src/common/verbose.hpp | aux stuff for MKLDNN_VERBOSE | |
| src/cpu/cpu_engine.cpp | the list of implementations | entry point for every implementation should be listed here so that the machinery will be able to pick it |
| src/cpu/cpu_softmax_pd.hpp | inherited class from softmax primitive descriptor that has some cpu specifics | and this specifics are the cpu_memory sources and desctinations |
| src/cpu/ref_softmax.cpp | a class with (reference) implementation | 
| src/cpu/ref_softmax.hpp | a class with (reference) implementation | exactly this class is included/registred in cpu_engine (more precisely one of its static function) |
| tests/gtests/CMakeLists.txt | Test list |  |
| tests/gtests/test_softmax_forward.cpp | Test for forward pass |
| tests/gtests/test_softmax_backward.cpp | Test for backward pass | Sometimes can be merged with the test for forward pass |

So if you decide to go with your implementation of ShuffleChannles primitive you most like will modify or extend the files in similar way softmax does. I will more than glad to help you with that if you wish. We can collaborate on your fork of Intel MKL-DNN, for example. Please also feel free to ask questions, as the code is not that easy to read.

I will also try to come up with some more thorough description of how Intel MKL-DNN machinery works, what are the main concepts and what are the typical steps to contribute to the library.

---

If you grep pooling (`grep pool`), you will notice the list of the files are almost the same as for softmax except for the bunch of implementations: 
```
src/cpu/jit_avx512_core_i8i8_pooling.hpp
src/cpu/jit_avx512_core_i8i8_pooling.cpp
src/cpu/jit_uni_pooling.hpp
src/cpu/jit_uni_pooling.cpp
src/cpu/jit_uni_pool_kernel_f32.hpp
src/cpu/jit_uni_pool_kernel_f32.cpp
src/cpu/nchw_pooling.hpp
src/cpu/nchw_pooling.cpp
src/cpu/ref_pooling.hpp
src/cpu/ref_pooling.cpp
```

Here `jit_*` files are the most optimized implementations we have, though they work for some particular data layout (called blocked: `nChw16c` and `nChw8c`). This layout is the main layout for Intel MKL-DNN since it allows you better utilize the vectorization capabilities of the hardware. Exactly due to this data layout we have so complicated logic of layout management (layout any, conversion from and to native data layout, etc...). And exactly for that format we need to have shuffle_channels_primitive cause the output of the convolution will most likely be `nChw16c` or `nChw8c`.
Here `nchw_*` files are the specialized implementation for `nchw` data layout. This implementation might be slower than the jitted one but it definitely much faster than the `ref_` one. The latter uses generic approach with fixed loop order and single data offset calculator (`memory_desc_wrapper::off(x0, x1, x2, ..., xn)`) which makes it horribly slow, though supporting all possible data layouts and data types.

**Comment by vpirogov** [comment]
Dropped the `help wanted` label, as we started looking at the design and implementation of shuffle primitive.

**Comment by vpirogov** [comment]
Implemented in d1e0b03
